concent_api/verifier/tasks.py

```diff
@@ -24,28 +35,95 @@ def blender_verification_order(
 
 
 @shared_task
+@transaction.atomic(using='control')
 def verification_result(
     subtask_id:     str,
     result:         VerificationResult,
-    error_message:  str,
-    error_code:     str,
+    error_message:  Optional[str] = None,
+    error_code:     Optional[ErrorCode]= None,
 ):
-    logger.info(
-        'verification_result_task starts with: SUBTASK_ID {} -- RESULT {}'.format(
-            subtask_id,
-            result,
-        )
-    )
+    logger.info(f'verification_result_task starts with: SUBTASK_ID {subtask_id} -- RESULT {result}')
 
     assert isinstance(subtask_id, str)
-    assert isinstance(error_message, str)
-    assert isinstance(error_code, str)
-    assert result in VerificationResult
+    assert isinstance(result, VerificationResult)
+    assert isinstance(error_message, (str, type(None)))
+    assert isinstance(error_code, (ErrorCode, type(None)))
     assert all([error_message, error_code]) if result == VerificationResult.ERROR else True
 
-    logger.info(
-        'verification_result_task ends with: SUBTASK_ID {} -- RESULT {}'.format(
-            subtask_id,
-            result,
+    # Worker locks database row corresponding to the subtask in the subtask table.
+    try:
+        subtask = Subtask.objects.select_for_update(nowait=True).get(subtask_id=subtask_id)
+    except DatabaseError:
+        # If the row is already locked, task fails so that Celery can retry later.
+        verification_result.delay(
+            subtask_id=subtask_id,
+            result=result,
+            error_message=error_message,
+            error_code=error_code,
+            countdown=CELERY_LOCKED_SUBTASK_DELAY
         )
-    )
+        return
+
+    if subtask.state_enum == Subtask.SubtaskState.ACCEPTED:
+        logger.warning(
+            'This means that the verification has timed out and a client has already asked about the result. '
+            'Or that verification_result for this subtask must have been already processed.'
+        )
+        return
+
+    elif subtask.state_enum == Subtask.SubtaskState.FAILED:
+        logger.error('This means that verification_result for this subtask must have been already processed.')
```

Comment:
Shouldn't this be `logger.warning()` as well? FAILED SubtaskState can happen because of timeout, which is not really an error.

Reply:
No. Timeout means that provider wins by default. So it goes into `ACCEPTED`.

Reply:
In issue

> B) If subtask is already in FAILED state, worker logs an error and returns. This means that verification_result for this subtask must have been already processed.

So I understood that I should 'log error'. But as you guys think, should it be changed ? Same question to 

```
    elif subtask.state_enum != Subtask.SubtaskState.ADDITIONAL_VERIFICATION:
        logger.error('This means that something is seriously wrong.')
```

Reply:
After thinking through I will change to log.warning if state is FAILED and keep log.error if state is unexpected.

Reply:
:+1:  done